### PR TITLE
License collection

### PR DESCRIPTION
**Pull request by estroz**
**PR details**: `collect_project_licenses.sh` downloads all dependencies of projects [listed](https://docs.google.com/spreadsheets/d/1gJIAHgEB8Av1gFNWznsx0nI_GHAEoN4gsiOc1NskJNc/edit#gid=0) which have 'Needs license info' == 'yes', finds their license info, and aggregates JSON output into a single list in a file. Go dependency licenses collected using [license-bill-of-materials](https://github.com/coreos/license-bill-of-materials); JS dependency license info collected using `package.json` files post-installation.

**Tests**: I tested the script locally. Will run in any environment that has Go, `yarn`, `jq`, and the `license-bill-of-materials` binary installed.
**Sample output**:
```
[
  {
    "project": "github.com/docker/distribution",
    "licenses": [
      {
        "type": "Apache License 2.0",
        "confidence": 1
      }
    ]
  },
  {
    "project": "github.com/docker/spdystream",
    "licenses": [
      {
        "type": "Apache License 2.0",
        "confidence": 0.9680365296803652
      },
      {
        "type": "Open Software License 3.0",
        "confidence": 0.4606413994169096
      }
    ]
  },
  ...
]
```

### CODE REVIEW

File: installer/scripts/release/collect_project_licenses.sh [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    if [ ! -f "$final_dep_file" ]; then
+      cat "$tmp_dep_file" > "$final_dep_file"
+    else
+      concat_deps_json "$tmp_dep_file" "$final_dep_file"
+    fi
+  done < "$entrypoints"
+}
+
+# Select relevant python binary
+this_python=python
```
Comment (sym3tri) [inline installer/scripts/release/collect_project_licenses.sh]: why do you need python? I'd rather we not introduce a new language dependency

File: installer/scripts/release/collect_project_licenses.sh [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  rm -rf "$new_gopath"
+  export GOPATH="$old_gopath"
+  exit
+}
+trap clean_up SIGHUP SIGINT SIGTERM EXIT
+
+set -eu
+# Grab all dependencies and install under tmp_gopath
+pushd . > /dev/null
+pkg_dirs=( "$TECTONIC_RELEASE_TOP_DIR" "$MATCHBOX_ARCHIVE_TOP_DIR" )
```
Comment (sym3tri) [inline installer/scripts/release/collect_project_licenses.sh]: where do these come from?

File: installer/scripts/release/collect_project_licenses.sh [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  rm -f "$final_dep"
+fi
+# Ensure GOPATH is set
+: "${GOPATH:?GOPATH is required but not set.}"
+# Ensure the license extractor binary is available
+if [ ! -f "$GOPATH"/bin/license-bill-of-materials ]; then
+  pushd . > /dev/null
+  cd "$GOPATH"
+  go get github.com/coreos/license-bill-of-materials
+  cd src/github.com/coreos/license-bill-of-materials
```
Comment (sym3tri) [inline installer/scripts/release/collect_project_licenses.sh]: You can just add this to our build container and assume it exists in the PATH
see: https://github.com/coreos/tectonic-installer/tree/master/images/tectonic-builder

File: installer/scripts/release/collect_project_licenses.sh (hunk 2 above)
Comment (sym3tri) [inline installer/scripts/release/collect_project_licenses.sh]: I'd ignore these and just use a prepopulated list of all the repos i sent you

File: installer/scripts/release/collect_project_licenses.sh [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+trap clean_up SIGHUP SIGINT SIGTERM EXIT
+
+set -eu
+# Grab all dependencies and install under tmp_gopath
+pushd . > /dev/null
+pkg_dirs=( "$TECTONIC_RELEASE_TOP_DIR" "$MATCHBOX_ARCHIVE_TOP_DIR" )
+pkg_main_locs=$(mktemp -p "$TMP_DIR")
+for d in "${pkg_dirs[@]}"; do
+  cd "$d"
+  url_prefix=$(pwd | grep -Eo "[^\/]*\.(com|org|net)\/.*")
```
Comment (sym3tri) [inline installer/scripts/release/collect_project_licenses.sh]: this seems fragile

**Comment by sym3tri** [comment]
@estroz can you include the output in this too?

File: installer/scripts/license-gen/collect_project_licenses.sh [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  # Create JSON objects using string concatenation
+  # TODO: remove duplicates (?)
+  tmp_json_arr=$(mktemp -p "$TMP_DIR")
+  trap "rm -f $tmp_json_arr; exit" SIGHUP SIGINT SIGTERM EXIT
+  # For every package.json found, parse out repository and license info
+  while read -r dep; do
+    obj=$(cat "$dep" | jq '{
+      project: (.repository | (.url)?//.),
+      licenses:
+        (if (.licenses)? then
```
Comment (rithujohn191) [inline installer/scripts/license-gen/collect_project_licenses.sh]: Please add a comment to describe what is going on in the if-else condition.

File: installer/scripts/license-gen/collect_project_licenses.sh [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+  find . \
+    -name main.go \
+    -not -path '*test*' \
+    -not -path '*vendor*' \
+    -not -path '*workspace*' \
+    -printf '%h\n' | uniq > "$1"
+  sed -i 's@^\.\(\/\)\?@@g' "$1"
+}
+
+set -eu
```
Comment (rithujohn191) [inline installer/scripts/license-gen/collect_project_licenses.sh]: Any reason why this cant be declared at the top and left off for the whole file?

File: installer/scripts/license-gen/collect_project_licenses.sh [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+  # Parse all package.json files for project and license fields
+  parse_js_licenses "$tmp_shared_file" "$js_pkg_json_file"
+done < "$js_pkg_inputs"
+popd > /dev/null
+
+# Append all licenses into one file
+while read -r file; do
+  # Concatenate all JSON arrays in a license file (might be 2-3 from
+  # license-bill-of-materials output)
+  curr_lic_json=$(cat "$file" | jq '.[]' | jq -s '.')
```
Comment (rithujohn191) [inline installer/scripts/license-gen/collect_project_licenses.sh]: Please rename this variable. json_licenses maybe?

**Pull request merged**